### PR TITLE
Refactor scaffolder to operate on known directories

### PR DESCRIPTION
**Pull request by jhaals**
## Hey, I just made a Pull Request!

- Generate a unique workspace directory inside the temp folder which gets cleaned up afterwards
- Refactor prepare/template/publish steps to operate on known directories(`checkout`, `template`, `result`) inside the temp directory. 
- Update preparers to take the template url instead of the entire template.

This is primarily steps required to allow for a smooth v1->v2 migration of templates.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

File: plugins/scaffolder-backend/src/scaffolder/stages/prepare/azure.test.ts [hunk 1]
```diff
@@ -36,119 +29,69 @@ describe('AzurePreparer', () => {
...
+        path,
+      )}`,
+      logger,
+      workspacePath: '/tmp',
     });
 
-    expect(response.split('\\').join('/')).toMatch(
-      /\/workDir\/graphql-starter-static\/template\/test\/1\/2\/3$/,
+    expect(fs.move).toHaveBeenCalledWith(
+      '/tmp/checkout/template/test/1/2/3',
```
Comment (OrkoHunter) [inline plugins/scaffolder-backend/src/scaffolder/stages/prepare/azure.test.ts:93]: Tests on Windows have been mysterious with mounting file system at `/tmp`.

I think it's safer to initialize an OS aware root folder like

```ts
const tmpDir = os.platform() === 'win32' ? 'C:\\tmp' : '/tmp';
```

and then use `resolve` from `'path'` to construct a paths inside `tmpDir`.

A recent related PR https://github.com/backstage/backstage/pull/4249

Comment (jhaals) [reply plugins/scaffolder-backend/src/scaffolder/stages/prepare/azure.test.ts:93]: Great suggestion, thanks! 👍

Comment (jhaals) [reply plugins/scaffolder-backend/src/scaffolder/stages/prepare/azure.test.ts]: @OrkoHunter I made some modifications to the test, please take a look before I replicate it to the rest of the preparers.

Comment (OrkoHunter) [reply plugins/scaffolder-backend/src/scaffolder/stages/prepare/azure.test.ts]: Your latest commit looks good to me @jhaals ! I have a trial of Windows VM via [parallels](https://www.parallels.com/) and so I took the liberty to run tests on this branch. Hope this log helps 😁 

https://pastebin.com/uQD1PG7a

I only ran the tests inside `plugins/scaffolder-backend`.

Comment (jhaals) [reply plugins/scaffolder-backend/src/scaffolder/stages/prepare/azure.test.ts]: Thanks a lot! I'm happy to see ` PASS  src/scaffolder/stages/prepare/azure.test.ts` :)

File: .changeset/wicked-beds-buy.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+'@backstage/plugin-scaffolder-backend': minor
+---
+
+The scaffolder is updated to generate a unique workspace directory inside the temp folder which gets cleaned up afterwards.
+
+prepare/template/publish steps is refactored to operate on known directories(`checkout/`, `template/`, `result/`) inside the generated temp directory.
+Updates preparers to take the template url instead of the entire template. This is all in preparation f
```
Comment (OrkoHunter) [inline .changeset/wicked-beds-buy.md:8]: f

Comment (jhaals) [reply .changeset/wicked-beds-buy.md:8]: for the great surprise* 😅

Comment (OrkoHunter) [reply .changeset/wicked-beds-buy.md:8]: Haha 😂

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4270-rxglpy1anpoqkc4jz9jxs2zv0gvugzmx.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6012cb9298552d29b3513f49

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4270-rxglpy1anpoqkc4jz9jxs2zv0gvugzmx.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6012cb9298552d29b3513f49

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4270-rxglpy1anpoqkc4jz9jxs2zv0gvugzmx.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6012cb9298552d29b3513f49

File: .changeset/wicked-beds-buy.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+---
+'@backstage/plugin-scaffolder-backend': minor
+---
+
+The scaffolder is updated to generate a unique workspace directory inside the temp folder which gets cleaned up afterwards.
+
+prepare/template/publish steps is refactored to operate on known directories(`checkout/`, `template/`, `result/`) inside the generated temp directory.
+Updates preparers to take the template url instead of the entire template. This is done primarily to allow for backwards compatibility between v1 and v2 scaffolder templates.
+
+Fix broken configuration GitHub actions in Create React App template.
```
Comment (Rugvip) [inline .changeset/wicked-beds-buy.md:10]:
```suggestion
Fix broken GitHub actions configuration in the Create React App template.
```

File: plugins/scaffolder-backend/src/scaffolder/stages/prepare/types.ts [hunk 4]
```diff
@@ -13,24 +13,25 @@
...
-   * Given an Entity definition from the Service Catalog, go and prepare a directory
-   * with contents from the remote location in temporary storage and return the path
-   * @param template The template entity from the Service Catalog
+   * Prepare a directory with contents from the remote location
    */
-  prepare(
-    template: TemplateEntityV1alpha1,
-    opts?: PreparerOptions,
-  ): Promise<string>;
+  prepare(opts: PreparerOptions): Promise<void>;
```
Comment (Rugvip) [inline plugins/scaffolder-backend/src/scaffolder/stages/prepare/types.ts:34]: Need to document this in the changeset along with a guide for how to migrate

File: plugins/scaffolder-backend/src/scaffolder/stages/publish/types.ts [hunk 5]
```diff
@@ -32,7 +32,7 @@ export type PublisherBase = {
 
 export type PublisherOptions = {
   values: TemplaterValues;
-  directory: string;
+  workspacePath: string;
```
Comment (Rugvip) [inline plugins/scaffolder-backend/src/scaffolder/stages/publish/types.ts:35]: Need a note in the changeset about this one too

File: plugins/scaffolder-backend/src/scaffolder/stages/templater/types.ts [hunk 6]
```diff
@@ -46,15 +45,15 @@ export type TemplaterRunResult = {
  * client to run any templater on top of your directory.
  */
 export type TemplaterRunOptions = {
-  directory: string;
+  workspacePath: string;
```
Comment (Rugvip) [inline plugins/scaffolder-backend/src/scaffolder/stages/templater/types.ts:48]: Changelog entry for this

File: plugins/scaffolder-backend/src/scaffolder/stages/templater/types.ts [hunk 7]
```diff
@@ -71,5 +70,5 @@ export type SupportedTemplatingKey = 'cookiecutter' | string;
  */
 export type TemplaterBuilder = {
   register(protocol: SupportedTemplatingKey, templater: TemplaterBase): void;
-  get(template: TemplateEntityV1alpha1): TemplaterBase;
+  get(templater: string): TemplaterBase;
```
Comment (Rugvip) [inline plugins/scaffolder-backend/src/scaffolder/stages/templater/types.ts:73]: And changelog entry for this 😁

**Review by Rugvip — COMMENTED** [review]
👍 , let's just get the changeset fill in a bit more :grin:

**Review by Rugvip — APPROVED** [review]

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4270-rxglpy1anpoqkc4jz9jxs2zv0gvugzmx.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6012cb9298552d29b3513f49

**Pull request merged**